### PR TITLE
FIX: castingHelper failed to find many_many_extraFields data (fixes #4661)

### LINKED ISSUE #4661: many_many_extraFields are not cast correctly in templates
They are always cast to whatever `default_cast` is set to. Easiest fix is probably to add a `DataObject::castingHelper()` method that looks for casting in extra fields.

### PR DESCRIPTION
**Pull request by kinglozzer**
Rebase of https://github.com/silverstripe/silverstripe-framework/pull/4667, fixes #4661.

### CODE REVIEW

File: model/DataObject.php [hunk 1]
```diff
@@ -1786,6 +1786,12 @@ public function getManyManyComponents($componentName, $filter = null, $sort = nu
 		$extraFields = $this->manyManyExtraFieldsForComponent($componentName) ?: array();
 		$result = ManyManyList::create($componentClass, $table, $componentField, $parentField, $extraFields);
 
+
+		// Store component data in query meta-data
+		$result = $result->alterDataQuery(function($query) use ($extraFields) {
+			$query->setQueryParam('Component.ExtraFields', $extraFields);
```
Comment (kinglozzer) [inline model/DataObject.php]: @tractorcow am I abusing this API again? :)

Comment (tractorcow) [inline model/DataObject.php]: Seems fine! :D

**Comment by tractorcow** [comment]
This seems like a pretty good way to get casting helpers across. :D Will wait for tests.

**Comment by kinglozzer** [comment]
Apparently I’m not worthy of a Travis build :( will try closing and re-opening to trigger one

**Pull request merged**